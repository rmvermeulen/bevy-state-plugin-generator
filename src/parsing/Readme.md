# parsing

Parsing is implemented using [`nom`](crates.io/nom). The result of a successfully
parsed text is the root `Node` which contains any children.

See the [tests](./tests.rs) for detailed examples of parsing behavior.
