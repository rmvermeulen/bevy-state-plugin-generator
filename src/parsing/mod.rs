pub(crate) mod header;
mod parsers;
#[cfg(test)]
mod tests;
mod tokens;

pub use parsers::*;
pub(crate) use tokens::*;

// TODO: investigate "tree-less parsing", see https://www.youtube.com/watch?v=NxiKlnUtyio
