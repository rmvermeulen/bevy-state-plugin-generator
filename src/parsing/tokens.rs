use derive_more::{Deref, From};

#[derive(Debug, PartialEq)]
pub enum Token {
    Separator,
    OpenEnum,
    CloseEnum,
    OpenList,
    CloseList,
}

#[derive(Clone, Copy, Debug, Deref, From, PartialEq)]
pub struct Identifier<'a>(&'a str);

impl std::fmt::Display for Identifier<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

#[derive(Clone, Copy, Debug, Deref, From, PartialEq)]
pub struct Comment<'a>(&'a str);

#[derive(Clone, Debug, PartialEq)]
pub enum ParseNode<'a> {
    Singleton(Identifier<'a>),
    Enum(Identifier<'a>, Vec<ParseNode<'a>>),
    List(Identifier<'a>, Vec<ParseNode<'a>>),
    Comment(Comment<'a>),
}

impl<'a> ParseNode<'a> {
    pub fn singleton<I: Into<Identifier<'a>>>(name: I) -> Self {
        Self::Singleton(name.into())
    }
    fn identifier(&self) -> Option<&Identifier<'a>> {
        match self {
            Self::Singleton(id) | Self::Enum(id, _) | Self::List(id, _) => Some(id),
            _ => None,
        }
    }
    pub fn name(&self) -> Option<&str> {
        self.identifier().map(|id| id.0)
    }
    pub fn children(&self) -> Vec<ParseNode<'a>> {
        match self {
            Self::Singleton(_) | Self::Comment(_) => Vec::new(),
            Self::Enum(_, children) => children.clone(),
            Self::List(_, children) => children.clone(),
        }
    }
    pub fn get_tree_size(&self) -> usize {
        match self {
            Self::Comment(_) => 1,
            Self::Singleton(_) => 1,
            Self::Enum(_, children) | Self::List(_, children) => {
                children
                    .iter()
                    .map(|child| child.get_tree_size())
                    .sum::<usize>()
                    + 1
            }
        }
    }
}

impl<'a> TryFrom<&'a str> for ParseNode<'a> {
    type Error = nom::Err<nom::error::Error<&'a str>>;

    fn try_from(s: &'a str) -> Result<Self, Self::Error> {
        crate::parsing::parse_node(s).map(|(_, node)| node)
    }
}

/// constructors used for testing
#[cfg(test)]
#[cfg_attr(coverage_nightly, coverage(off))]
impl<'a> ParseNode<'a> {
    pub fn enumeration<I: Into<Identifier<'a>>, V: IntoIterator<Item = ParseNode<'a>>>(
        name: I,
        variants: V,
    ) -> Self {
        Self::Enum(name.into(), variants.into_iter().collect())
    }
    pub fn list_empty<I: Into<Identifier<'a>>>(name: I) -> Self {
        Self::List(name.into(), vec![])
    }
    pub fn list<I: Into<Identifier<'a>>, V: IntoIterator<Item = ParseNode<'a>>>(
        name: I,
        variants: V,
    ) -> Self {
        Self::List(name.into(), variants.into_iter().collect())
    }
    pub fn comment<C: Into<Comment<'a>>>(name: C) -> Self {
        Self::Comment(name.into())
    }
}
