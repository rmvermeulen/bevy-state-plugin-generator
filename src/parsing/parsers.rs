use nom::branch::alt;
use nom::bytes::complete::*;
use nom::character::complete::*;
use nom::combinator::recognize;
use nom::multi::many0;
use nom::sequence::*;
use nom::{IResult, Parser};

use crate::parsing::{Comment, Identifier, ParseNode, Token};

pub fn parse_comment(input: &str) -> IResult<&str, ParseNode<'_>> {
    comment(input).map_result(ParseNode::Comment)
}

pub fn comment(input: &str) -> IResult<&str, Comment<'_>> {
    use nom::combinator::eof;

    delimited(skip_to(tag("//")), not_line_ending, alt((eof, line_ending)))
        .parse(input)
        .map_result(|c| c.trim().into())
}

pub fn identifier(input: &str) -> IResult<&str, Identifier<'_>> {
    recognize(pair(
        take_till1(|c: char| !c.is_uppercase()),
        take_while(|c: char| c.is_alphanumeric() || c == '_'),
    ))
    .parse(input)
    .map_result(Identifier::from)
}

fn skip_to<P, T, O, E>(parser: P) -> impl Parser<T, Output = O, Error = E>
where
    T: nom::Input,
    P: Parser<T, Output = O, Error = E>,
    <T as nom::Input>::Item: nom::AsChar,
    E: nom::error::ParseError<T>,
{
    preceded(multispace0, parser)
}

pub fn separator(input: &str) -> IResult<&str, Token> {
    skip_to(tag(","))
        .parse(input)
        .map_result(|_| Token::Separator)
}

pub fn open_enum(input: &str) -> IResult<&str, Token> {
    skip_to(tag("{"))
        .parse(input)
        .map_result(|_| Token::OpenEnum)
}

pub fn close_enum(input: &str) -> IResult<&str, Token> {
    skip_to(tag("}"))
        .parse(input)
        .map_result(|_| Token::CloseEnum)
}

pub fn open_list(input: &str) -> IResult<&str, Token> {
    skip_to(tag("["))
        .parse(input)
        .map_result(|_| Token::OpenList)
}

pub fn close_list(input: &str) -> IResult<&str, Token> {
    skip_to(tag("]"))
        .parse(input)
        .map_result(|_| Token::CloseList)
}

pub trait MapResult<'a, I, O1, O2> {
    fn map_result(self, callback: impl Fn(O1) -> O2) -> IResult<I, O2>;
}

impl<I, O1, O2> MapResult<'_, I, O1, O2> for IResult<I, O1> {
    fn map_result(self, callback: impl Fn(O1) -> O2) -> IResult<I, O2> {
        self.map(|(rest, result)| (rest, callback(result)))
    }
}

pub fn parse_config(input: &str) -> IResult<&str, Vec<ParseNode<'_>>> {
    many0(delimited(many0(separator), parse_node, many0(separator))).parse(input)
}

pub fn parse_node(input: &'_ str) -> IResult<&'_ str, ParseNode<'_>> {
    alt((parse_enum, parse_list, parse_comment, parse_singleton)).parse(input)
}

pub fn parse_singleton(input: &'_ str) -> IResult<&'_ str, ParseNode<'_>> {
    skip_to(identifier)
        .parse(input)
        .map_result(ParseNode::singleton)
}

pub fn parse_enum(input: &str) -> IResult<&str, ParseNode<'_>> {
    let (input, name) = terminated(skip_to(identifier), skip_to(open_enum)).parse(input)?;
    let (input, children) = skip_to(parse_elements_until(close_enum)).parse(input)?;
    Ok((input, ParseNode::Enum(name, children)))
}

pub fn parse_list(input: &'_ str) -> IResult<&'_ str, ParseNode<'_>> {
    let (input, name) = skip_to(identifier).parse(input)?;
    let (input, children) =
        skip_to(preceded(open_list, parse_elements_until(close_list))).parse(input)?;
    Ok((input, ParseNode::List(name, children)))
}

pub fn parse_elements_until<'a>(
    until: impl Fn(&'a str) -> IResult<&'a str, Token> + Copy,
) -> impl Fn(&'a str) -> IResult<&'a str, Vec<ParseNode<'a>>> {
    move |input: &'a str| {
        delimited(
            // ignore any leading whitespace and commas
            many0(separator),
            // 0 or more elements, ignoring whitespace and commas
            many0(alt((terminated(parse_node, many0(separator)),))),
            // then expect the closing token
            until,
        )
        .parse(input)
    }
}
