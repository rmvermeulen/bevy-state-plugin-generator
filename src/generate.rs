#[cfg(test)]
mod tests;

use indoc::formatdoc;
use itertools::Itertools;
use nom::AsChar;

use crate::PluginConfig;
use crate::parsing::parse_states_text;
use crate::processing::{ProcessingError, try_parse_node_into_final_source};

pub(super) const REQUIRED_DERIVES: &[&str] =
    &["Hash", "Default", "Debug", "Clone", "PartialEq", "Eq"];

pub fn get_package_info() -> String {
    let pkg = env!("CARGO_PKG_NAME");
    #[cfg(not(test))]
    let version = env!("CARGO_PKG_VERSION");
    #[cfg(test)]
    let version = "[CARGO_PKG_VERSION]";
    format!("{pkg} v{version}")
}

pub fn generate_debug_info(src_path: &str, source: &str) -> String {
    let lines = source.lines().map(|line| format!("// {line}")).join("\n");
    let pkg_info = get_package_info();
    formatdoc! {"
        // generated by {pkg_info}
        // src: {src_path}
        {lines}
    "}
}

#[cfg(feature = "rustfmt")]
fn try_format_source(source: &str) -> std::io::Result<String> {
    duct::cmd!("rustfmt")
        .stdin_bytes(source)
        .stderr_to_stdout()
        .read()
}

pub(crate) fn format_source<S: AsRef<str>>(source: S) -> String {
    let source = source.as_ref();
    #[cfg(feature = "rustfmt")]
    let source = try_format_source(source).unwrap_or_else(|_| source.to_owned());
    #[cfg(not(feature = "rustfmt"))]
    let source = source.to_owned();

    if source.ends_with(|c: char| c.is_newline()) {
        source
    } else {
        source + "\n"
    }
}

pub fn generate_state_plugin_source(
    input_source: &str,
    plugin_config: PluginConfig,
    src_path: Option<&str>,
) -> Result<String, ProcessingError> {
    let parse_nodes = parse_states_text(input_source)?;
    let mut output = parse_nodes
        .into_iter()
        .map(|parse_node| try_parse_node_into_final_source(parse_node, plugin_config.clone()))
        .collect::<Result<Vec<_>, _>>()?
        .join("\n");
    if let Some(src_path) = src_path {
        let debug_info = generate_debug_info(src_path, input_source);
        output = [debug_info, output].join("\n");
    }
    Ok(format_source(output))
}
