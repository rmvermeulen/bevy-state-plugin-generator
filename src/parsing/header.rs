use itertools::Itertools;
use lazy_regex::regex;

use crate::generate::get_package_info;
use crate::{NamingScheme, PluginConfig};

#[derive(Clone, Debug, Default, PartialEq)]
pub(crate) struct TemplateHeader<'a> {
    pub template: Vec<&'a str>,
    pub comments_block: Vec<&'a str>,
    pub info_block: Vec<String>,
}

trait Unquote {
    fn unquoted(&self) -> Self;
}

impl Unquote for &str {
    fn unquoted(&self) -> Self {
        self.strip_prefix("\"")
            .and_then(|s| s.strip_suffix("\""))
            .unwrap_or(self)
    }
}

pub(crate) fn parse_template_header<'a>(
    source: &'a str,
    plugin_config: &mut PluginConfig,
) -> TemplateHeader<'a> {
    let leading_comments = source
        .lines()
        .take_while(|line| line.starts_with("//"))
        .collect_vec();

    let mut info_block = Vec::new();
    info_block.push(format!("generated by {}", get_package_info()));

    macro_rules! warn {
        ($($exprs:expr),*) => {
            let warning = format!($($exprs),*);
            let warning = format!("WARN: {warning}");
            eprintln!("{warning}");
            info_block.push(warning);
        };
    }

    let mut template_source = Vec::new();
    let mut in_template = false;
    let leading_comments = leading_comments
        .iter()
        .filter(|line| {
            if in_template {
                if let Some(line) = line.strip_prefix("//") {
                    template_source.push(line.trim());
                    true
                } else {
                    false
                }
            } else if let Some(captures) =
                regex!(r#"^\s*//\s*bspg:(\w+)\s+(\w+|"\w+\")\s*$"#).captures(line)
            {
                let (_, [name, value]) = captures.extract();
                match name {
                    "root_state_name" => {
                        plugin_config.root_state_name = if value == "None" {
                            None
                        } else {
                            Some(value.unquoted().to_string())
                        };
                    }
                    "naming_scheme" => {
                        if let Some(naming_scheme) = NamingScheme::try_parse(value) {
                            plugin_config.naming_scheme = naming_scheme;
                        } else {
                            warn!("invalid naming scheme '{value}' (expected [none, short, full])");
                        }
                    }
                    _ => {
                        warn!("unknown setting: '{name}'");
                    }
                }
                true
            } else if regex!(r#"^\s*//\s*bspg:\s*$"#).is_match(line) {
                in_template = true;
                true
            } else {
                eprintln!("dropping: '{line:?}'");
                false
            }
        })
        .copied()
        .collect_vec();
    TemplateHeader {
        template: template_source,
        comments_block: leading_comments,
        info_block,
    }
}

#[cfg(test)]
mod tests {
    use bevy_utils::default;
    use indoc::formatdoc;
    use rstest::rstest;
    use speculoos::assert_that;
    use speculoos::prelude::{ContainingIntoIterAssertions, VecAssertions};

    use crate::parsing::header::parse_template_header;
    use crate::{NamingScheme, PluginConfig};

    #[rstest]
    #[case(String::new())]
    #[case::ignore_generated_lines(formatdoc! {"
        // generated by v[CARGO_PKG_VERSION]
        // WARN: some warning
    "})]
    fn test_parse_template_header(#[case] header: String) {
        let mut config: PluginConfig = default();
        let header = parse_template_header(&header, &mut config);
        assert_that!(header.template).is_empty();
        assert_that!(header.comments_block).is_empty();
    }

    #[rstest]
    fn test_parse_template_header_keep_variables() {
        let header = formatdoc! {"
            // generated by v[CARGO_PKG_VERSION]
            // WARN: some warning
            // bspg:some setting
        "};
        let mut config: PluginConfig = default();
        let header = parse_template_header(&header, &mut config);
        assert_that!(header.comments_block).contains("// bspg:some setting");
        assert_that!(header.info_block).contains("WARN: unknown setting: 'some'".to_string());
    }

    #[rstest]
    fn test_parse_template_header_variable_values(
        #[values(NamingScheme::Full, NamingScheme::None, NamingScheme::Short)]
        naming_scheme: NamingScheme,
        #[values(None, Some("MyStateRoot"))] name: Option<&str>,
    ) {
        let mut config: PluginConfig = default();
        let header = formatdoc! {"
            // generated by v[CARGO_PKG_VERSION]
            // bspg:root_state_name {}
            // bspg:naming_scheme   {}
            // bspg:
            ",
            name.map(|name| format!("\"{name}\"")).unwrap_or("None".to_string()),
            naming_scheme
        };
        let _ = parse_template_header(&header, &mut config);
        assert_that!(config.root_state_name).is_equal_to(name.map(String::from));
        assert_that!(config.naming_scheme).is_equal_to(naming_scheme);
    }
}
