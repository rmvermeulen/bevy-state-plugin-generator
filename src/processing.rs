#[cfg(test)]
mod tests;

use std::collections::VecDeque;

use bevy_platform::collections::HashSet;
use bevy_utils::default;
use indoc::formatdoc;
use itertools::{Itertools, concat};

use crate::generate::REQUIRED_DERIVES;
use crate::parsing::ParseNode;
use crate::{NamingScheme, PluginConfig, PluginName};

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub enum NodeType {
    #[default]
    Singleton,
    List,
    Enum,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct NodeData {
    pub index: usize,
    pub parent: Option<usize>,
    pub node_type: NodeType,
    pub depth: usize,
    pub name: String,
    pub resolved_name: Option<String>,
    pub variants: Vec<String>,
}

pub fn flatten_root_parse_node(root_node: ParseNode<'_>) -> Vec<NodeData> {
    let node_count = root_node.get_tree_size();
    let mut nodes = Vec::with_capacity(node_count);
    let mut todo = VecDeque::from([(root_node, 0, None)]);
    while let Some((parse_node, depth, parent)) = todo.pop_front() {
        let node_type = match parse_node {
            ParseNode::Singleton(_) => NodeType::Singleton,
            ParseNode::Enum(_, _) => NodeType::Enum,
            ParseNode::List(_, _) => NodeType::List,
            ParseNode::Comment(_) => {
                continue;
            }
        };
        let Some(name) = parse_node.name() else {
            continue;
        };
        let index = nodes.len();
        nodes.push(NodeData {
            node_type,
            index,
            parent,
            depth,
            name: name.to_string(),
            ..default()
        });
        for child in parse_node.children() {
            todo.push_back((child, depth + 1, Some(index)));
        }
    }

    for i in 0..nodes.len() {
        let (ancestors, children) = nodes.split_at_mut(i);
        let node = &children[0];
        if let Some(parent_index) = node.parent {
            assert!(parent_index < i);
            let parent = &mut ancestors[parent_index];
            parent.variants.push(node.name.clone());
        }
    }

    #[cfg(debug_assertions)]
    for (i, node) in nodes.iter().enumerate() {
        assert_eq!(node.index, i);
        if let Some(parent_index) = node.parent {
            assert!(parent_index <= node.index);
        }
    }

    nodes
}

pub fn apply_naming_scheme(
    naming_scheme: NamingScheme,
    nodes: &mut [NodeData],
) -> Result<(), ProcessingError> {
    let mut names = HashSet::new();
    let mut resolved_names = Vec::new();
    for node in nodes.iter() {
        let base_name = &node.name;
        let resolved_name = match naming_scheme {
            NamingScheme::Full => {
                let mut ancestors = Vec::new();
                let mut current = node;
                while let Some(p) = current.parent {
                    current = &nodes[p];
                    ancestors.push(current);
                }
                let ancestral_name = ancestors.into_iter().map(|a| &a.name).rev().join("");
                format!("{ancestral_name}{base_name}")
            }
            NamingScheme::Short => {
                let parent = node
                    .parent
                    .map(|p| nodes[p].name.as_str())
                    .unwrap_or_default();
                format!("{parent}{base_name}")
            }
            NamingScheme::None => base_name.clone(),
        };
        if !names.insert(resolved_name.clone()) {
            return Err(ProcessingError::DuplicateName {
                resolved_name,
                original_name: base_name.clone(),
            });
        }

        resolved_names.push(resolved_name);
    }
    assert_eq!(names.len(), resolved_names.len());
    for (i, resolved_name) in resolved_names.into_iter().enumerate() {
        nodes[i].resolved_name = Some(resolved_name);
    }
    Ok(())
}

type NomError<T> = nom::Err<nom::error::Error<T>>;

#[derive(Debug, thiserror::Error)]
pub enum ProcessingError {
    #[error("Duplicate name: resolved_name='{resolved_name}' original_name='{original_name}'")]
    DuplicateName {
        resolved_name: String,
        original_name: String,
    },
    #[error("Unspecified error: {0}")]
    Custom(String),
    #[error("Failed to parse! Final state: {0:?}")]
    Parsing(#[from] nom::Err<nom::error::Error<String>>),
    #[error("Invalid configuration: {0}")]
    InvalidConfig(String),
}

impl From<NomError<&str>> for ProcessingError {
    fn from(value: NomError<&str>) -> Self {
        Self::from(value.to_owned())
    }
}

fn build_plugin_source(
    nodes: Vec<NodeData>,
    config: PluginConfig,
) -> Result<String, ProcessingError> {
    let PluginConfig {
        additional_derives: derives,
        plugin_name,
        root_state_name,
        naming_scheme: _,
        states_module_name,
    } = config;

    let derives = concat([
        REQUIRED_DERIVES
            .iter()
            .map(ToString::to_string)
            .collect_vec(),
        derives.into_iter().collect_vec(),
    ])
    .into_iter()
    .unique()
    .join(", ");
    let definitions_source = nodes
        .iter()
        .map(|node| {
            let derives = node
                .parent
                .map(|parent_id| {
                    let parent = &nodes[parent_id];
                    formatdoc! {"
                        #[derive(bevy::prelude::SubStates, {derives})]
                        #[source({source} = {source}::{variant})]
                    ", variant = node.name, source = parent.name
                    }
                })
                .unwrap_or_else(|| formatdoc! {"#[derive(bevy::prelude::States, {derives})]"})
                .trim()
                .to_string();
            let typename = &node.name;

            let source_for_singleton = || {
                formatdoc! {"
                {derives}
                pub struct {typename};
            "}
            };
            let source_for_enum = |variants: &[&str]| {
                formatdoc! {"
                {derives}
                pub enum {typename} {{
                    #[default]
                    {variants}
                }}
                ", variants = variants.join(",\n      ")
                }
            };

            match node.node_type {
                NodeType::Singleton => source_for_singleton(),
                NodeType::Enum => {
                    if node.variants.is_empty() {
                        source_for_singleton()
                    } else {
                        let variants = node.variants.iter().map(String::as_str).collect_vec();
                        source_for_enum(&variants)
                    }
                }
                NodeType::List => source_for_singleton(),
            }
        })
        .join("\n");

    let plugin_builder = if let Some(root_state_name) = root_state_name.as_ref() {
        let states_module_name = states_module_name.as_str();
        let init_state = format!(".init_state::<{states_module_name}::{root_state_name}>()");
        let sub_states = nodes
            .iter()
            .flat_map(|node| {
                let resolved_name = node
                    .resolved_name
                    .as_ref()
                    .expect("Node name has not been resolved!");
                if node.parent.map(|p| nodes[p].clone()).is_some() {
                    Some(format!(
                        ".add_sub_state::<{states_module_name}::{resolved_name}>()"
                    ))
                } else {
                    None
                }
            })
            .join("\n            ");
        format!("app{init_state}{sub_states};")
    } else {
        let states_module_name = states_module_name.as_str();
        let states = nodes
            .iter()
            .map(|node| {
                let resolved_name = node
                    .resolved_name
                    .clone()
                    .expect("Node name has not been resolved!");
                if node.parent.map(|p| nodes[p].clone()).is_some() {
                    format!(".add_sub_state::<{states_module_name}::{resolved_name}>()")
                } else {
                    format!(".add_state::<{states_module_name}::{resolved_name}>()")
                }
            })
            .join("\n            ");
        format!("app{states};")
    };

    let plugin_def = match plugin_name {
        PluginName::Struct(plugin_name) => {
            formatdoc! {"
                pub struct {plugin_name};
                impl bevy::app::Plugin for {plugin_name} {{
                    fn build(&self, app: &mut bevy::app::App) {{
                        {plugin_builder}
                    }}
                }}
            "}
        }
        PluginName::Function(plugin_name) => {
            formatdoc! {"
                pub fn {plugin_name}(app: &mut bevy::app::App) {{
                    {plugin_builder}
                }}
            "}
        }
    };

    Ok(formatdoc! {"
        use bevy::prelude::AppExtStates;
        #[allow(missing_docs)]
        pub mod {states_module_name} {{
            use bevy::prelude::StateSet;
            {definitions_source}
        }}
        {plugin_def}
    "})
}

pub(crate) fn remove_root_node(nodes: &mut Vec<NodeData>) {
    nodes.remove(0);
    for node in nodes {
        assert!(node.index > 0);
        node.index -= 1;
        assert!(node.depth > 0);
        node.depth -= 1;
        if let Some(parent_index) = &mut node.parent {
            node.parent = (*parent_index > 0).then(|| *parent_index - 1);
        }
    }
}

pub(crate) fn process_parse_nodes(
    parse_nodes: Vec<ParseNode<'_>>,
    naming_scheme: NamingScheme,
    root_state_name: Option<String>,
) -> Result<Vec<NodeData>, ProcessingError> {
    // add the implicit root_node according to config
    let root_node = if let Some(root_state_name) = &root_state_name {
        ParseNode::Enum(root_state_name.as_str().into(), parse_nodes)
    } else {
        // add a temporary root node
        ParseNode::List("".into(), parse_nodes)
    };
    let mut nodes = flatten_root_parse_node(root_node);
    // remove temporary root node
    if root_state_name.is_none() {
        remove_root_node(&mut nodes);
    }
    apply_naming_scheme(naming_scheme, &mut nodes)?;
    Ok(nodes)
}

pub fn convert_parse_nodes_into_plugin_source(
    parse_nodes: Vec<ParseNode<'_>>,
    config: PluginConfig,
) -> Result<String, ProcessingError> {
    let nodes = process_parse_nodes(
        parse_nodes,
        config.naming_scheme,
        config.root_state_name.clone(),
    )?;
    assert!(nodes.iter().all(|node| node.resolved_name.is_some()));
    build_plugin_source(nodes, config)
}
