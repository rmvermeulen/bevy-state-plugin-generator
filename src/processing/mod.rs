#[cfg(test)]
mod tests;

use std::collections::VecDeque;

use bevy_platform::collections::HashSet;
use bevy_utils::default;
use indoc::formatdoc;
use itertools::{Itertools, concat};

use crate::generate::REQUIRED_DERIVES;
use crate::parsing::ParseNode;
use crate::tree::SubTree;
use crate::{NamingScheme, PluginConfig, PluginName};

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub enum NodeType {
    #[default]
    Singleton,
    List,
    Enum,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct NodeData {
    pub index: usize,
    pub parent: Option<usize>,
    pub node_type: NodeType,
    pub depth: usize,
    pub name: String,
    pub resolved_name: Option<String>,
    pub variants: Vec<String>,
}

pub fn flatten_parse_node(root_node: ParseNode<'_>) -> Vec<NodeData> {
    let node_count = root_node.get_tree_size();
    let mut nodes = Vec::with_capacity(node_count);
    let mut todo = VecDeque::from([(root_node, 0, None)]);
    while let Some((parse_node, depth, parent)) = todo.pop_front() {
        let Some(name) = parse_node.name() else {
            continue;
        };
        let node_type = match parse_node {
            ParseNode::Singleton(_) => NodeType::Singleton,
            ParseNode::Enum(_, _) => NodeType::Enum,
            ParseNode::List(_, _) => NodeType::List,
            ParseNode::Comment(_) => {
                unreachable!("Comment has no name")
            }
        };
        let index = nodes.len();
        nodes.push(NodeData {
            node_type,
            index,
            parent,
            depth,
            name: name.to_string(),
            ..default()
        });
        for child in parse_node.children() {
            todo.push_back((child, depth + 1, Some(index)));
        }
    }

    for (i, node) in nodes.iter().enumerate() {
        assert_eq!(node.index, i);
        if let Some(parent) = node.parent {
            assert!(parent <= node.index);
            assert!(node.depth > nodes[parent].depth);
        }
    }

    nodes
}

pub fn apply_naming_scheme(
    naming_scheme: NamingScheme,
    nodes: &mut [NodeData],
) -> Result<(), ProcessingError> {
    let mut names = HashSet::new();
    let mut resolved_names = Vec::new();
    for node in nodes.iter() {
        let base_name = &node.name;
        let resolved_name = match naming_scheme {
            NamingScheme::Full => {
                let mut ancestors = Vec::new();
                let mut current = node;
                while let Some(p) = current.parent {
                    current = &nodes[p];
                    ancestors.push(current);
                }
                let ancestral_name = ancestors.into_iter().map(|a| &a.name).join("");
                format!("{ancestral_name}{base_name}")
            }
            NamingScheme::Short => {
                let parent = node
                    .parent
                    .map(|p| nodes[p].name.as_str())
                    .unwrap_or_default();
                format!("{parent}{base_name}")
            }
            NamingScheme::None => base_name.clone(),
        };
        if !names.insert(resolved_name.clone()) {
            return Err(ProcessingError::DuplicateName {
                resolved_name,
                original_name: base_name.clone(),
            });
        }

        resolved_names.push(resolved_name);
    }
    assert_eq!(names.len(), resolved_names.len());
    for (i, resolved_name) in resolved_names.into_iter().enumerate() {
        nodes[i].resolved_name = Some(resolved_name);
    }
    Ok(())
}

#[derive(Debug, thiserror::Error)]
pub enum ProcessingError {
    #[error("Duplicate name: resolved_name='{resolved_name}' original_name='{original_name}'")]
    DuplicateName {
        resolved_name: String,
        original_name: String,
    },
    #[error("Unspecified error: {0}")]
    Custom(String),
    #[error("Failed to parse! Final state: {0:?}")]
    Parsing(String),
}

fn get_source(nodes: Vec<NodeData>, config: PluginConfig) -> Result<String, ProcessingError> {
    let PluginConfig {
        additional_derives: derives,
        plugin_name,
        root_state_name,
        naming_scheme: _,
        states_module_name,
    } = config;

    // add the implicit root_node according to config
    let nodes = if let Some(root_state_name) = &root_state_name {
        let root_node = NodeData {
            index: 0,
            parent: None,
            name: root_state_name.clone(),
            resolved_name: Some(root_state_name.clone()),
            ..default()
        };
        let mut new_nodes = vec![root_node];
        for mut node in nodes.into_iter() {
            let root_node = &mut new_nodes[0];
            if node.depth == 0 {
                root_node.variants.push(
                    node.resolved_name
                        .clone()
                        .unwrap_or_else(|| panic!("Node name has not been resolved! {node:?}")),
                );
            }
            // adjust indices, since we've added a root_node
            node.index += 1;
            node.depth += 1;
            if let Some(parent_index) = &mut node.parent {
                *parent_index += 1;
            } else {
                // parent this node to the root_node
                _ = node.parent.insert(0);
            }
            assert!(node.parent.is_some());
            new_nodes.push(node);
        }
        new_nodes
    } else {
        nodes
    };

    let derives = concat([
        REQUIRED_DERIVES
            .iter()
            .map(ToString::to_string)
            .collect_vec(),
        derives.into_iter().collect_vec(),
    ])
    .into_iter()
    .unique()
    .join(", ");
    let definitions_source = nodes
        .iter()
        .map(|node| {
            let derives = node
                .parent
                .map(|parent_id| {
                    let parent = &nodes[parent_id];
                    formatdoc! {"
                        #[derive(bevy::prelude::SubStates, {derives})]
                        #[source({source} = {source}::{variant})]
                    ", variant = node.name, source = parent.name
                    }
                })
                .unwrap_or_else(|| formatdoc! {"#[derive(bevy::prelude::States, {derives})]"})
                .trim()
                .to_string();
            let typename = &node.name;

            let source_for_singleton = || {
                formatdoc! {"
                {derives}
                pub struct {typename};
            "}
            };
            let source_for_enum = |variants: &[&str]| {
                formatdoc! {"
                {derives}
                pub enum {typename} {{
                    #[default]
                    {variants}
                }}
                ", variants = variants.join(",\n      ")
                }
            };

            match node.node_type {
                NodeType::Singleton => source_for_singleton(),
                NodeType::Enum => {
                    if node.variants.is_empty() {
                        source_for_singleton()
                    } else {
                        let variants = node.variants.iter().map(String::as_str).collect_vec();
                        source_for_enum(&variants)
                    }
                }
                NodeType::List => source_for_singleton(),
            }
        })
        .join("\n");

    let plugin_builder = if let Some(root_state_name) = root_state_name.as_ref() {
        let states_module_name = states_module_name.as_str();
        let init_state = format!(".init_state::<{states_module_name}::{root_state_name}>()");
        let sub_states = nodes
            .iter()
            .flat_map(|node| {
                let resolved_name = node
                    .resolved_name
                    .as_ref()
                    .expect("Node name has not been resolved!");
                if node.parent.map(|p| nodes[p].clone()).is_some() {
                    Some(format!(
                        ".add_sub_state::<{states_module_name}::{resolved_name}>()"
                    ))
                } else {
                    None
                }
            })
            .join("\n            ");
        format!("app{init_state}{sub_states};")
    } else {
        let states_module_name = states_module_name.as_str();
        let states = nodes
            .iter()
            .map(|node| {
                let resolved_name = node
                    .resolved_name
                    .clone()
                    .expect("Node name has not been resolved!");
                if node.parent.map(|p| nodes[p].clone()).is_some() {
                    format!(".add_sub_state::<{states_module_name}::{resolved_name}>()")
                } else {
                    format!(".add_state::<{states_module_name}::{resolved_name}>()")
                }
            })
            .join("\n            ");
        format!("app{states};")
    };

    let plugin_def = match plugin_name {
        PluginName::Struct(plugin_name) => {
            formatdoc! {"
                pub struct {plugin_name};
                impl bevy::app::Plugin for {plugin_name} {{
                    fn build(&self, app: &mut bevy::app::App) {{
                        {plugin_builder}
                    }}
                }}
            "}
        }
        PluginName::Function(plugin_name) => {
            formatdoc! {"
                pub fn {plugin_name}(app: &mut bevy::app::App) {{
                    {plugin_builder}
                }}
            "}
        }
    };

    Ok(formatdoc! {"
        #![allow(missing_docs)]
        use bevy::prelude::AppExtStates;
        pub mod {states_module_name} {{
            use bevy::prelude::StateSet;
            {definitions_source}
        }}
        {plugin_def}
    "})
}

pub fn parse_node_into_final_source(
    node: ParseNode<'_>,
    config: PluginConfig,
) -> Result<String, ProcessingError> {
    let mut nodes = flatten_parse_node(node);
    apply_naming_scheme(config.naming_scheme, &mut nodes)?;
    assert!(nodes.iter().all(|node| node.resolved_name.is_some()));
    get_source(nodes, config)
}
