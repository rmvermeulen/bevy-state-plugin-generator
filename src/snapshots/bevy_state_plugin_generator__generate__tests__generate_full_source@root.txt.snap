---
source: src/generate.rs
expression: "generate_full_source(src_path, source, plugin_config).unwrap_or_else(identity)"
---
// src: root.txt
// RootState

#![allow(missing_docs)]
use bevy::prelude::AppExtStates;

pub mod states {
    use bevy::prelude::StateSet;
    #[derive(bevy::prelude::States, Hash, Default, Debug, Clone, PartialEq, Eq)]
    pub enum GameState {
        #[default]
        RootState,
    }

    #[derive(bevy::prelude::SubStates, Hash, Default, Debug, Clone, PartialEq, Eq)]
    #[source(GameStateRootState = GameStateRootState::RootState)]
    pub struct GameStateRootStateRootState;
}
pub struct GeneratedStatesPlugin;
impl bevy::app::Plugin for GeneratedStatesPlugin {
    fn build(&self, app: &mut bevy::app::App) {
        app.init_state::<states::GameState>();
    }
}
